### PR TITLE
Some cleaning of Database error dialog

### PR DESCRIPTION
**Pull request by Arthur-Milchior**
While working on #17399 I realized that similar cleaning could occur to the nullable deck picker that can't actually be null.

### CODE REVIEW

**Review by mikehardy — APPROVED** [review]
There likely won't be any *more* ClassCastExceptions from this but we may still have some.

I was just in here fixing some ClassCastException crashes but those were related to import and I made the required type there sufficiently generic (AnkiActivity) that it works from the two Activities that spawn import processes

The rest of the activities, if they need deck picker but are spawned from some other activity will just keep crashing until the features they need are also moved up to a common superclass or pulled out to some object that's usable from any activity

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt [hunk 1]
```diff
@@ -84,6 +84,8 @@ import java.io.IOException
 class DatabaseErrorDialog : AsyncDialogFragment() {
     private lateinit var backups: Array<File>
 
+    private fun requireDeckPicker() = activity as DeckPicker
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt:87]: Separates the two types of exceptions which can occur

```suggestion
    private fun requireDeckPicker() = requireActivity() as DeckPicker
```

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt [hunk 2]
```diff
@@ -110,8 +112,7 @@ class DatabaseErrorDialog : AsyncDialogFragment() {
                     message(text = message)
                     setIcon(R.drawable.ic_warning)
                     positiveButton(R.string.error_handling_options) {
-                        (activity as DeckPicker?)
-                            ?.showDatabaseErrorDialog(DIALOG_ERROR_HANDLING, exceptionData)
+                        requireDeckPicker().showDatabaseErrorDialog(DIALOG_ERROR_HANDLING, exceptionData)
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt:115]: We should always pass in `exceptionData` here, so:

```patch
Index: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt b/AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt
--- a/AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt	(revision a2b6699d30340939bf478e4453a156e18d3cfd4f)
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt	(date 1733617497999)
@@ -112,7 +112,7 @@
                     message(text = message)
                     setIcon(R.drawable.ic_warning)
                     positiveButton(R.string.error_handling_options) {
-                        requireDeckPicker().showDatabaseErrorDialog(DIALOG_ERROR_HANDLING, exceptionData)
+                        showDatabaseErrorDialog(DIALOG_ERROR_HANDLING)
                     }
                     negativeButton(R.string.close) {
                         closeCollectionAndFinish()
@@ -128,7 +128,7 @@
                     message(text = message)
                     setIcon(R.drawable.ic_warning)
                     positiveButton(R.string.error_handling_options) {
-                        requireDeckPicker().showDatabaseErrorDialog(DIALOG_ERROR_HANDLING, exceptionData)
+                        showDatabaseErrorDialog(DIALOG_ERROR_HANDLING)
                     }
                     negativeButton(R.string.answering_error_report) {
                         requireDeckPicker().sendErrorReport()
@@ -183,19 +183,19 @@
                                 return@listItems
                             }
                             ErrorHandlingEntries.REPAIR -> {
-                                requireDeckPicker().showDatabaseErrorDialog(DIALOG_REPAIR_COLLECTION, exceptionData)
+                                showDatabaseErrorDialog(DIALOG_REPAIR_COLLECTION)
                                 return@listItems
                             }
                             ErrorHandlingEntries.RESTORE -> {
-                                requireDeckPicker().showDatabaseErrorDialog(DIALOG_RESTORE_BACKUP, exceptionData)
+                                showDatabaseErrorDialog(DIALOG_RESTORE_BACKUP)
                                 return@listItems
                             }
                             ErrorHandlingEntries.ONE_WAY -> {
-                                requireDeckPicker().showDatabaseErrorDialog(DIALOG_ONE_WAY_SYNC_FROM_SERVER, exceptionData)
+                                showDatabaseErrorDialog(DIALOG_ONE_WAY_SYNC_FROM_SERVER)
                                 return@listItems
                             }
                             ErrorHandlingEntries.NEW -> {
-                                requireDeckPicker().showDatabaseErrorDialog(DIALOG_NEW_COLLECTION, exceptionData)
+                                showDatabaseErrorDialog(DIALOG_NEW_COLLECTION)
                                 return@listItems
                             }
 
@@ -233,7 +233,7 @@
                     alertDialog.title(R.string.backup_restore)
                         .title(text = message)
                         .positiveButton(R.string.dialog_ok) {
-                            requireDeckPicker().showDatabaseErrorDialog(DIALOG_ERROR_HANDLING, exceptionData)
+                            showDatabaseErrorDialog(DIALOG_ERROR_HANDLING)
                         }
                 } else {
                     // Show backups sorted with latest on top
@@ -290,7 +290,7 @@
                         if (BackupManager.moveDatabaseToBrokenDirectory(path1, false, time)) {
                             ActivityCompat.recreate(requireDeckPicker())
                         } else {
-                            requireDeckPicker().showDatabaseErrorDialog(DIALOG_LOAD_FAILED, exceptionData)
+                            showDatabaseErrorDialog(DIALOG_LOAD_FAILED)
                         }
                     }
                     negativeButton(R.string.dialog_cancel)
@@ -314,7 +314,7 @@
                     title(R.string.restore_backup_title)
                     message(text = message)
                     positiveButton(R.string.dialog_continue) {
-                        requireDeckPicker().showDatabaseErrorDialog(DIALOG_RESTORE_BACKUP, exceptionData)
+                        showDatabaseErrorDialog(DIALOG_RESTORE_BACKUP)
                     }
                     negativeButton(R.string.dialog_cancel)
                 }
@@ -360,13 +360,11 @@
                     }
                     listItemsAndMessage(message = message, items = options) { _, index: Int ->
                         when (values[index]) {
-                            IncompatibleDbVersionEntries.RESTORE -> requireDeckPicker().showDatabaseErrorDialog(
-                                DIALOG_RESTORE_BACKUP,
-                                exceptionData
+                            IncompatibleDbVersionEntries.RESTORE -> showDatabaseErrorDialog(
+                                DIALOG_RESTORE_BACKUP
                             )
-                            IncompatibleDbVersionEntries.ONE_WAY -> requireDeckPicker().showDatabaseErrorDialog(
-                                DIALOG_ONE_WAY_SYNC_FROM_SERVER,
-                                exceptionData
+                            IncompatibleDbVersionEntries.ONE_WAY -> showDatabaseErrorDialog(
+                                DIALOG_ONE_WAY_SYNC_FROM_SERVER
                             )
                         }
                     }
@@ -398,6 +396,11 @@
         }
     }
 
+    /** @see DeckPicker.showDatabaseErrorDialog */
+    private fun showDatabaseErrorDialog(errorDialogType: DatabaseErrorDialogType) {
+        requireDeckPicker().showDatabaseErrorDialog(errorDialogType, exceptionData)
+    }
+
     /**
      * Collects the current exception's stack trace and debug information,
      * and copies the combines information to the Android clipboard.

```

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt [hunk 3]
```diff
@@ -179,23 +179,23 @@ class DatabaseErrorDialog : AsyncDialogFragment() {
                     listItems(items = options) { _, index ->
                         when (values[index]) {
                             ErrorHandlingEntries.RETRY -> {
-                                ActivityCompat.recreate(activity as DeckPicker)
+                                ActivityCompat.recreate(requireDeckPicker())
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt:182]: Is this necessary? Can we use `requireActivity()`

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt [hunk 4]
```diff
@@ -290,9 +288,9 @@ class DatabaseErrorDialog : AsyncDialogFragment() {
                         CollectionManager.closeCollectionBlocking()
                         val path1 = CollectionHelper.getCollectionPath(requireActivity())
                         if (BackupManager.moveDatabaseToBrokenDirectory(path1, false, time)) {
-                            ActivityCompat.recreate(activity as DeckPicker)
+                            ActivityCompat.recreate(requireDeckPicker())
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt:291]: ditto

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt [hunk 5]
```diff
@@ -390,7 +387,7 @@ class DatabaseErrorDialog : AsyncDialogFragment() {
                     title(R.string.directory_inaccessible_after_uninstall)
                     listItemsAndMessage(message = message, items = listItems.map { getString(it.stringRes) }) { _, index: Int ->
                         val listItem = listItems[index]
-                        listItem.onClick(activity as DeckPicker)
+                        listItem.onClick(requireDeckPicker())
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt:390]: ditto: can this be `requireActivity`

**Review by david-allison — APPROVED** [review]
nitpicks only. Here is an untested patch containing them all

```patch
Index: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt b/AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt
--- a/AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt	(revision a2b6699d30340939bf478e4453a156e18d3cfd4f)
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt	(date 1733617572218)
@@ -56,6 +56,7 @@
 import com.ichi2.anki.dialogs.ImportFileSelectionFragment.ImportOptions
 import com.ichi2.anki.isLoggedIn
 import com.ichi2.anki.launchCatchingTask
+import com.ichi2.anki.requireAnkiActivity
 import com.ichi2.anki.servicelayer.DebugInfoService
 import com.ichi2.anki.showImportDialog
 import com.ichi2.anki.utils.ext.dismissAllDialogFragments
@@ -84,7 +85,7 @@
 class DatabaseErrorDialog : AsyncDialogFragment() {
     private lateinit var backups: Array<File>
 
-    private fun requireDeckPicker() = activity as DeckPicker
+    private fun requireDeckPicker() = requireActivity() as DeckPicker
 
     @SuppressLint("CheckResult")
     override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
@@ -112,7 +113,7 @@
                     message(text = message)
                     setIcon(R.drawable.ic_warning)
                     positiveButton(R.string.error_handling_options) {
-                        requireDeckPicker().showDatabaseErrorDialog(DIALOG_ERROR_HANDLING, exceptionData)
+                        showDatabaseErrorDialog(DIALOG_ERROR_HANDLING)
                     }
                     negativeButton(R.string.close) {
                         closeCollectionAndFinish()
@@ -128,7 +129,7 @@
                     message(text = message)
                     setIcon(R.drawable.ic_warning)
                     positiveButton(R.string.error_handling_options) {
-                        requireDeckPicker().showDatabaseErrorDialog(DIALOG_ERROR_HANDLING, exceptionData)
+                        showDatabaseErrorDialog(DIALOG_ERROR_HANDLING)
                     }
                     negativeButton(R.string.answering_error_report) {
                         requireDeckPicker().sendErrorReport()
@@ -179,23 +180,23 @@
                     listItems(items = options) { _, index ->
                         when (values[index]) {
                             ErrorHandlingEntries.RETRY -> {
-                                ActivityCompat.recreate(requireDeckPicker())
+                                ActivityCompat.recreate(requireActivity())
                                 return@listItems
                             }
                             ErrorHandlingEntries.REPAIR -> {
-                                requireDeckPicker().showDatabaseErrorDialog(DIALOG_REPAIR_COLLECTION, exceptionData)
+                                showDatabaseErrorDialog(DIALOG_REPAIR_COLLECTION)
                                 return@listItems
                             }
                             ErrorHandlingEntries.RESTORE -> {
-                                requireDeckPicker().showDatabaseErrorDialog(DIALOG_RESTORE_BACKUP, exceptionData)
+                                showDatabaseErrorDialog(DIALOG_RESTORE_BACKUP)
                                 return@listItems
                             }
                             ErrorHandlingEntries.ONE_WAY -> {
-                                requireDeckPicker().showDatabaseErrorDialog(DIALOG_ONE_WAY_SYNC_FROM_SERVER, exceptionData)
+                                showDatabaseErrorDialog(DIALOG_ONE_WAY_SYNC_FROM_SERVER)
                                 return@listItems
                             }
                             ErrorHandlingEntries.NEW -> {
-                                requireDeckPicker().showDatabaseErrorDialog(DIALOG_NEW_COLLECTION, exceptionData)
+                                showDatabaseErrorDialog(DIALOG_NEW_COLLECTION)
                                 return@listItems
                             }
 
@@ -233,7 +234,7 @@
                     alertDialog.title(R.string.backup_restore)
                         .title(text = message)
                         .positiveButton(R.string.dialog_ok) {
-                            requireDeckPicker().showDatabaseErrorDialog(DIALOG_ERROR_HANDLING, exceptionData)
+                            showDatabaseErrorDialog(DIALOG_ERROR_HANDLING)
                         }
                 } else {
                     // Show backups sorted with latest on top
@@ -288,9 +289,9 @@
                         CollectionManager.closeCollectionBlocking()
                         val path1 = CollectionHelper.getCollectionPath(requireActivity())
                         if (BackupManager.moveDatabaseToBrokenDirectory(path1, false, time)) {
-                            ActivityCompat.recreate(requireDeckPicker())
+                            ActivityCompat.recreate(requireActivity())
                         } else {
-                            requireDeckPicker().showDatabaseErrorDialog(DIALOG_LOAD_FAILED, exceptionData)
+                            showDatabaseErrorDialog(DIALOG_LOAD_FAILED)
                         }
                     }
                     negativeButton(R.string.dialog_cancel)
@@ -314,7 +315,7 @@
                     title(R.string.restore_backup_title)
                     message(text = message)
                     positiveButton(R.string.dialog_continue) {
-                        requireDeckPicker().showDatabaseErrorDialog(DIALOG_RESTORE_BACKUP, exceptionData)
+                        showDatabaseErrorDialog(DIALOG_RESTORE_BACKUP)
                     }
                     negativeButton(R.string.dialog_cancel)
                 }
@@ -360,13 +361,11 @@
                     }
                     listItemsAndMessage(message = message, items = options) { _, index: Int ->
                         when (values[index]) {
-                            IncompatibleDbVersionEntries.RESTORE -> requireDeckPicker().showDatabaseErrorDialog(
-                                DIALOG_RESTORE_BACKUP,
-                                exceptionData
+                            IncompatibleDbVersionEntries.RESTORE -> showDatabaseErrorDialog(
+                                DIALOG_RESTORE_BACKUP
                             )
-                            IncompatibleDbVersionEntries.ONE_WAY -> requireDeckPicker().showDatabaseErrorDialog(
-                                DIALOG_ONE_WAY_SYNC_FROM_SERVER,
-                                exceptionData
+                            IncompatibleDbVersionEntries.ONE_WAY -> showDatabaseErrorDialog(
+                                DIALOG_ONE_WAY_SYNC_FROM_SERVER
                             )
                         }
                     }
@@ -387,7 +386,7 @@
                     title(R.string.directory_inaccessible_after_uninstall)
                     listItemsAndMessage(message = message, items = listItems.map { getString(it.stringRes) }) { _, index: Int ->
                         val listItem = listItems[index]
-                        listItem.onClick(requireDeckPicker())
+                        listItem.onClick(requireAnkiActivity())
                         if (listItem.dismissesDialog) {
                             dismiss()
                         }
@@ -398,6 +397,11 @@
         }
     }
 
+    /** @see DeckPicker.showDatabaseErrorDialog */
+    private fun showDatabaseErrorDialog(errorDialogType: DatabaseErrorDialogType) {
+        requireDeckPicker().showDatabaseErrorDialog(errorDialogType, exceptionData)
+    }
+
     /**
      * Collects the current exception's stack trace and debug information,
      * and copies the combines information to the Android clipboard.
```

**Review by mikehardy — APPROVED** [review]
I think everything optional addressed now even

**Pull request merged**